### ISSUE TITLE
Support boolean indexing

### ISSUE THREAD

**Issue by dashesy** [onnx/onnx#2302]
Gather currently does not accept a bool tensor

**Comment by wschin** [comment]
For your example,
```
idx = a > 0.5
w[idx] = b[idx]
```
it seems you are looking for `Where` operator.

**Comment by dashesy** [comment]
Thanks, yes `where` for that and `nonzero` for `w = b[nonzero(a > 0.5).view(-1)]`. But still boolean indexing would be nice, specially because then I would not have to use something like `.view(-1)` when the tensors are more than 2 dimensional

**Comment by wschin** [comment]
> Thanks, yes where for that and nonzero for w = b[nonzero(a > 0.5).view(-1, 1)]. But still boolean indexing would be nice, specially because then I would not have to use something like .view(-1, 1) when the tensors are more than 2 dimensional.
> Boolean indexing can be broadcasted.

I am not sure why you have to use `.view(-1, 1)`. The selection mask of `Where` is broadcastable. Do you have another example for cases where `Where` doesn't work?

**Comment by dashesy** [comment]
I think you are right, so PyTorch exporter **can** use `where` ops when boolean indices are used.

**Comment by jinfagang** [comment]
How to using where to replace bool indexing then?

**Comment by jinfagang** [comment]
I don't think where can replace Gather (bool index). For example:

```
torch.where(condition, x, y) → Tensor
```

Returns a same size tensor as input, but Gather only gather these true values. Can they equal?
